manifest: Adding nRF Services library

Rafal-Nordic opened the pull request:
Adding nRF Services library to the hal-nordic repo

Depends on https://github.com/zephyrproject-rtos/docker-image/pull/181

Rafal-Nordic committed:
manifest: Adding nRF Services library

Adding nRF Services library to the hal-nordic repo

Signed-off-by: Rafal Dyla <rafal.dyla@nordicsemi.no>

jfischer-no commented on modules/hal_nordic/nrfs/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if(CONFIG_NRFS)
+    if(CONFIG_NRFS_LOCAL_DOMAIN)
```
CMakeLists.txt in the hal/nordic/nrfs is not included.
```diff
diff --git a/modules/hal_nordic/nrfs/CMakeLists.txt b/modules/hal_nordic/nrfs/CMakeLists.txt
index a4edc6be513..eec370cf9b1 100644
--- a/modules/hal_nordic/nrfs/CMakeLists.txt
+++ b/modules/hal_nordic/nrfs/CMakeLists.txt
@@ -2,6 +2,8 @@
 # SPDX-License-Identifier: Apache-2.0
 
 if(CONFIG_NRFS)
+    add_subdirectory(${ZEPHYR_CURRENT_MODULE_DIR}/nrfs nrfs)
+
     if(CONFIG_NRFS_LOCAL_DOMAIN)
         zephyr_sources_ifdef(CONFIG_NRFS_BACKEND_IPC_SERVICE_EXTENDED
                             backends/nrfs_backend_ipc_service.c)
```

Rafal-Nordic replied:
CMakeLists are now completely reorganized.

jfischer-no commented on modules/hal_nordic/nrfs/Kconfig:
```diff
@@ -0,0 +1,121 @@
+	depends on NRFS_HAS_TEMP_SERVICE
+	default y
+
+config USB_SERVICE_ENABLED
+	bool "USB service"
+	depends on NRFS_HAS_USB_SERVICE
+	default y
```
There must not be anything prefixed with USB_ or USB_SERVICE_ or have prompt "USB service" that is not part of the USB support in Zephyr. You need to find another short and concise name for it.

Rafal-Nordic replied:
done

jfischer-no commented on modules/hal_nordic/nrfs/Kconfig:
```diff
@@ -0,0 +1,121 @@
+	help
+	  This option enables the nRF Services library
+
+config NRFS_LOCAL_DOMAIN
```
What exactly is LOCAL_DOMAIN? Is there also REMOTE_DOMAIN?

hubertmis replied:
There are local and global domains in this SoC architecture

Rafal-Nordic replied:
"Local domain" name comes from Product Specification and is commonly used in the MDK. nrf54H20 has the following local domains:
- App core
- Rad core
- Secure domain

jfischer-no commented on modules/hal_nordic/nrfs/Kconfig:
```diff
@@ -0,0 +1,121 @@
+	help
+	  This option enables the nRF Services Local Domain libraries
+
+config NRFS_BACKEND_IPC_SERVICE
```
Can NRFS be use without IPC?

Rafal-Nordic replied:
SysCtrl supports only IPC_SERVICE backend protocol today, but other domains use also other protocols, so I don't exclude that in the future additional protocols will be supported.

jfischer-no commented on modules/hal_nordic/nrfs/backends/Kconfig:
```diff
@@ -0,0 +1,89 @@
+	depends on NRFS_BACKEND_IPC_SERVICE
+	default NRFS_BACKEND_IPC_SERVICE_LITE
+
+config NRFS_BACKEND_IPC_SERVICE_EXTENDED
+	bool "Zephyr IPC service backend for NRFS with additional buffering."
+
+config NRFS_BACKEND_IPC_SERVICE_LITE
+	bool "Use IPC service directly by NRFS"
```
Why two backends and what are the differences?

Rafal-Nordic replied:
I checked that "standard" version doesn't make sense anymore. I'll remove it

Rafal-Nordic replied:
removed

jfischer-no commented on modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.c:
```diff
@@ -0,0 +1,341 @@
+};
+
+static struct ipc_channel_config_t ipc_cpusys_channel_config = {
+	.ipc_instance	 = DEVICE_DT_GET(DT_NODELABEL(ipc_to_cpusys)),
```
That finally blocked me, I could not fix it myself quickly. Neither `ipc_to_cpusys` nor `cpuapp_cpusys_ipc` seems to exist with all the other things that belong to it. It is like a rocket without a propulsion system. Same for the lite backend.

Rafal-Nordic replied:
Yes, not all required dts files were moved to the upstream repo. I'm checking how to fix it

Rafal-Nordic replied:
Implemented here: https://github.com/zephyrproject-rtos/zephyr/pull/71265

tmon-nordic commented on modules/hal_nordic/nrfs/Kconfig:
```diff
@@ -0,0 +1,121 @@
+config NRFS_BACKEND_IPC_SERVICE
+	bool "IPC service backend for NRFS"
+	depends on NRFS
+	imply IPC_SERVICE
+	imply IPC_SERVICE_BACKEND_ICMSG
+	imply MBOX
+	imply EVENTS
+	imply REBOOT
```
Disabling the dependencies result in link failure and therefore these should be forcing dependencies (`select`), not suggestions (`imply`).

tmon-nordic commented on modules/hal_nordic/nrfs/Kconfig:
```diff
@@ -0,0 +1,121 @@
+	help
+	  This option enables the nRF Services Local Domain libraries
+
+config NRFS_BACKEND_IPC_SERVICE
+	bool "IPC service backend for NRFS"
+	depends on NRFS
+	select IPC_SERVICE
+	select IPC_SERVICE_BACKEND_ICMSG
+	select MBOX
+	select EVENTS
+	select REBOOT
+	help
+	  This option enables transport protocol for the NRFS
```
How does it enables the transport protocol? I cannot find any use of `CONFIG_NRFS_BACKEND_IPC_SERVICE`.

Rafal-Nordic replied:
fixed

tmon-nordic commented on modules/hal_nordic/nrfs/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+
+if(CONFIG_NRFS)
+    add_subdirectory(${ZEPHYR_CURRENT_MODULE_DIR}/nrfs nrfs)
+    if(CONFIG_NRFS_LOCAL_DOMAIN)
```
`nrfs_backend_ipc_service_lite.c` sounds like `CONFIG_NRFS_BACKEND_IPC_SERVICE` and not `CONFIG_NRFS_LOCAL_DOMAIN`. Not quite sure where is a problem, but it is `nrfs_backend_ipc_service_lite.c` that uses the `NRFS_BACKEND_IPC_SERVICE` dependencies.

Rafal-Nordic replied:
fixed

tmon-nordic reviewed (changes requested):
I think it would be worthwhile to add dependencies to devicetree in the Kconfigs, i.e.:
`depends on $(dt_nodelabel_has_compat,cpuapp_cpusys_ipc,$(DT_COMPAT_ZEPHYR_IPC_ICMSG))`

Rafal-Nordic commented on modules/hal_nordic/nrfs/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
```
CMakeLists are now completely reorganized.

tmon-nordic commented on modules/hal_nordic/nrfs/Kconfig:
```diff
@@ -0,0 +1,125 @@
+	bool "MRAM latency service"
+	depends on NRFS_HAS_MRAM_SERVICE
+	default y
+
+config TEMP_SERVICE_ENABLED
+	bool "Temperature service"
+	depends on NRFS_HAS_TEMP_SERVICE
+	default y
+
+config VBUS_DETECTOR_SERVICE_ENABLED
+	bool "VBUS detector for the USB peripheral"
+	depends on NRFS_HAS_VBUS_DETECTOR_SERVICE
+	default y
+
+config CONST_LATENCY_SERVICE_ENABLED
+	bool "DPPI constant latency service"
+	depends on NRFS_HAS_CONST_LATENCY_SERVICE
+	default y
+
+config PMIC_SERVICE_ENABLED
+	bool "PMIC service"
+	depends on NRFS_HAS_PMIC_SERVICE
+	default n
+
+config DVFS_SERVICE_ENABLED
+	bool "DVFS service"
+	depends on NRFS_HAS_DVFS_SERVICE
+	default y if SOC_NRF54H20_CPUAPP
+	default n
+
+config DIAG_SERVICE_ENABLED
+	bool "System Diagnostics service (only for development purposes)"
+	depends on NRFS_HAS_DIAG_SERVICE
+	default n
+
+config CLOCK_SERVICE_ENABLED
+	bool "Clock service"
+	depends on NRFS_HAS_CLOCK_SERVICE
+	default y
+endmenu
```
All the configs here should be namespaced, i.e. start with `NRFS_` prefix. Otherwise it can be really confusing what e.g. `CONFIG_CLOCK_SERVICE_ENABLED` refers to.

Rafal-Nordic replied:
right, I'll change it

tmon-nordic reviewed (dismissed):
`west build -p -b nrf54h20dk/nrf54h20/cpuapp -d build-54h20-hello samples/hello_world/ -DCONFIG_NRFS=y -DCONFIG_NRFS_LOCAL_DOMAIN=y -DCONFIG_NRFS_BACKEND_IPC_SERVICE=y`

fails with

```
warning: IPC_SERVICE_BACKEND_ICMSG (defined at subsys/ipc/ipc_service/backends/Kconfig:19) has direct dependencies MBOX && DT_HAS_ZEPHYR_IPC_ICMSG_ENABLED && IPC_SERVICE with value n, but is currently being y-selected by the following symbols:
- NRFS_BACKEND_IPC_SERVICE (defined at modules/hal_nordic/nrfs/Kconfig:50, modules/hal_nordic/nrfs/Kconfig:50), with value y, direct dependencies (NRFS && NRF_HAS_NRFS) || (NRFS && NRF_HAS_NRFS && 0) (value: y), and select condition NRFS && NRF_HAS_NRFS (value: y)



error: Aborting due to Kconfig warnings
```

It might also be worthwhile to add `depends on $(dt_alias_enabled,ipc-to-cpusys)` to `config NRFS_BACKEND_IPC_SERVICE`.

anangl commented on modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.c:
```diff
@@ -0,0 +1,225 @@
+
+#include <internal/nrfs_backend.h>
+#include <internal/nrfs_dispatcher.h>
+#include <nrfs_backend_ipc_service.h>
```
```suggestion
#include "nrfs_backend_ipc_service.h"
```

anangl commented on modules/hal_nordic/nrfs/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+            message(WARNING "This service is for Nordic Semiconductor INTERNAL purposes ONLY. Use it with caution due to risk of hardware damage!")
+        endif()
+    endif()
+    zephyr_sources_ifdef(CONFIG_NRFS_BACKEND_IPC_SERVICE backends/nrfs_backend_ipc_service.c)
```
Why not `zephyr_library_sources_ifdef` (also in `dvfs/CMakeLists.txt`)?

anangl commented on modules/hal_nordic/nrfs/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+        endif()
+    endif()
+    zephyr_sources_ifdef(CONFIG_NRFS_BACKEND_IPC_SERVICE backends/nrfs_backend_ipc_service.c)
+    add_subdirectory_ifdef(CONFIG_NRFS_DVFS_LOCAL_DOMAIN dvfs)
```
Shouldn't this be under `if(CONFIG_NRFS_LOCAL_DOMAIN)` above for consistency?

anangl commented on modules/hal_nordic/nrfs/Kconfig:
```diff
@@ -0,0 +1,126 @@
+	depends on NRFS
+	depends on $(dt_alias_enabled,ipc-to-cpusys)
+	select IPC_SERVICE
+	select IPC_SERVICE_BACKEND_ICMSG
```
Is it needed? The backend should be automatically enabled based on the enabled node in dts: https://github.com/zephyrproject-rtos/zephyr/blob/a6400b42feeb2cc44aacb14d7b29370602cb9ce2/subsys/ipc/ipc_service/backends/Kconfig#L19-L23

anangl commented on modules/hal_nordic/nrfs/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+
+    if(CONFIG_NRFS_LOCAL_DOMAIN)
+        zephyr_library_sources(${SRC_DIR}/internal/nrfs_dispatcher.c)
+        zephyr_library_sources_ifdef(CONFIG_NRFS_TEMP_SERVICE_ENABLED ${SRC_DIR}/services/nrfs_temp.c)
+        zephyr_library_sources_ifdef(CONFIG_NRFS_MRAM_SERVICE_ENABLED ${SRC_DIR}/services/nrfs_mram.c)
+        zephyr_library_sources_ifdef(CONFIG_NRFS_RESET_SERVICE_ENABLED ${SRC_DIR}/services/nrfs_reset.c)
+        zephyr_library_sources_ifdef(CONFIG_NRFS_VBUS_DETECTOR_SERVICE_ENABLED ${SRC_DIR}/services/nrfs_usb.c)
+        zephyr_library_sources_ifdef(CONFIG_NRFS_PMIC_SERVICE_ENABLED ${SRC_DIR}/services/nrfs_pmic.c)
+        zephyr_library_sources_ifdef(CONFIG_NRFS_DVFS_SERVICE_ENABLED ${SRC_DIR}/services/nrfs_dvfs.c)
+        zephyr_library_sources_ifdef(CONFIG_NRFS_DIAG_SERVICE_ENABLED ${SRC_DIR}/services/nrfs_diag.c)
+        zephyr_library_sources_ifdef(CONFIG_NRFS_CLOCK_SERVICE_ENABLED ${SRC_DIR}/services/nrfs_clock.c)
```
Why not have these entries ordered by name?

Rafal-Nordic commented on modules/hal_nordic/nrfs/Kconfig:
```diff
@@ -0,0 +1,125 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+config NRF_HAS_NRFS
```
I agree, it is a better name

anangl commented on boards/nordic/nrf54h20dk/nrf54h20dk_nrf54h20_cpuapp.dts:
```diff
@@ -131,6 +136,7 @@ ipc0: &cpuapp_cpurad_ipc {
 };
 
 &cpuapp_cpusys_ipc {
+	status = "okay";
 	mbox-names = "rx", "tx";
 	tx-region = <&cpuapp_cpusys_ipc_shm>;
 	rx-region = <&cpusys_cpuapp_ipc_shm>;
```
Also the `shared_ram20_region` node needs to be enabled as the above shared memory regions are located there.

anangl committed:
drivers: sensor: Add driver for nRF temperature sensor accessed via nrfs

Add driver, together with the corresponding dts binding and node in
the nRF54H20 SoC definiton, for the nRF temperature sensor that cannot
be accessed directly but only through nRF Services (nrfs) layer.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
tests: drivers: Add test for temperature sensor

Add a simple test intended primarily for checking the nRF temperature
sensor in its nrfs variant but intentionally was made generic enough
that it should be usable with other temperature sensors (after just
providing an overlay) and could be helpful in development of new
temperature sensor drivers, and possibly extended then if needed.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -0,0 +1,378 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(LD_DVFS_LIB, CONFIG_LOCAL_DOMAIN_DVFS_LIB_LOG_LEVEL);
+
+K_SEM_DEFINE(dvfs_service_sync_sem, 0, 1);
+K_SEM_DEFINE(dvfs_service_idle_sem, 0, 1);
+
+#define DVFS_SERV_HDL_INIT_DONE_BIT_POS (0)
+#define DVFS_SERV_HDL_FREQ_CHANGE_IN_PROGRESS_BIT_POS (1)
+
+atomic_t dvfs_service_handler_state_bits;
```
Is there a reason why these variables are not `static`?

anangl replied:
It looks like the semaphores above could be made `static`, too.

Rafal-Nordic replied:
done

tmon-nordic commented:
Please remove the cherry-picked from lines from commit messages. Cherry-picked from lines are useful for the backports to older releases or cherry-picks to downstream repositories, but here you are submitting brand new commits to upstream main.

tmon-nordic reviewed (changes requested):
Please remove the irrelevant cherry-picked from lines like explained earlier.

jfischer-no commented on modules/hal_nordic/nrfs/Kconfig:
```diff
@@ -0,0 +1,122 @@
+
+config NRFS_BACKEND_IPC_SERVICE
+	bool "IPC service backend for NRFS"
+	depends on $(dt_alias_enabled,ipc-to-cpusys)
```
This is redundant and already covered by IPC_SERVICE.

jfischer-no commented on modules/hal_nordic/nrfs/backends/Kconfig:
```diff
@@ -0,0 +1,28 @@
+module-str = NRFS backend
+source "subsys/logging/Kconfig.template.log_config"
+
+config NRFS_BACKEND_IPC_SERVICE_LITE_INIT_PRIO
```
Since there is only one backend, can _LITE be removed?

jfischer-no commented on modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.c:
```diff
@@ -0,0 +1,237 @@
+
+static void nrfs_backend_send_work(struct k_work *item)
+{
+	LOG_DBG("Sending data from workqueue");
+	static struct ipc_data_packet_t data_to_send;
+
+	while (k_msgq_get(&ipc_transmit_msgq, &data_to_send, K_NO_WAIT) == 0) {
```
Variable definition should be at the top of a block.
```c
	static struct ipc_data_packet_t data_to_send;

	LOG_DBG("Sending data from workqueue");
	while (k_msgq_get(&ipc_transmit_msgq, &data_to_send, K_NO_WAIT) == 0) {
```

jfischer-no commented on modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.c:
```diff
@@ -0,0 +1,237 @@
+ */
+static int ipc_channel_init(void)
+{
+	int ret = 0;
+
+	k_work_init(&backend_send_work, nrfs_backend_send_work);
+
+	struct ipc_channel_config_t *ipc_channel_config = &ipc_cpusys_channel_config;
```
No reason to avoid concise variable names, please place the variable definition at the top of a block.
```c
	struct ipc_channel_config_t *ch_cfg;
	int ret = 0;

	k_work_init(&backend_send_work, nrfs_backend_send_work);
	ch_cfg = &ipc_cpusys_channel_config;
```

jfischer-no commented on modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.c:
```diff
@@ -0,0 +1,237 @@
+	}
+}
+
+/**
+ *  @brief Initialize ipc channel
+ *
+ *  @param ipc_channel_config pointer to ipc channel config structure
+ *
+ *  @return -EINVAL when instance configuration is invalid.
+ *  @return -EIO when no backend is registered.
+ *  @return -EALREADY when the instance is already opened (or being opened).
+ *  @return -EBUSY when the instance is busy.
+ *  @return 0 on suscess
+ */
```
Doxygen style documentation of a static function is anything but helpful, how and where should it be rendered and why?

Rafal-Nordic replied:
It was only added to better describe return values. There was no intention to render this.

jfischer-no commented on modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.c:
```diff
@@ -0,0 +1,237 @@
+	}
+
+	if (size <= MAX_PACKET_DATA_SIZE) {
+		tx_data.channel_id = IPC_CPUSYS_CHANNEL_ID;
+		tx_data.size	   = size;
+		memcpy(tx_data.data, (uint8_t *)message, size);
+
+		int err = k_msgq_put(&ipc_transmit_msgq, &tx_data, timeout);
+
+		if (err) {
```
```c
		int err;

		tx_data.channel_id = IPC_CPUSYS_CHANNEL_ID;
		tx_data.size	   = size;
		memcpy(tx_data.data, (uint8_t *)message, size);

		err = k_msgq_put(&ipc_transmit_msgq, &tx_data, timeout);
		if (err) {
```

jfischer-no commented on modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.c:
```diff
@@ -0,0 +1,237 @@
+
+	LOG_ERR("Trying to send %d bytes where max is %d.", size, MAX_PACKET_DATA_SIZE);
+	return NRFS_ERR_IPC;
+}
```
```c

	LOG_ERR("Trying to send %d bytes where max is %d.", size, MAX_PACKET_DATA_SIZE);

	return NRFS_ERR_IPC;
}
```

jfischer-no commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -0,0 +1,378 @@
+
+static enum dvfs_frequency_setting current_freq_setting;
+
+/**
+ * @brief Helper function to set bit in state variable.
+ *
+ * @param bit_pos bit position in state variable.
+ */
```
Please remove redundant comments / redundant doxygen comments.

Rafal-Nordic replied:
I understand that documentation for static functions is not needed, but in the case we have it I would like to leave it as is.

jfischer-no replied:
Again, please remove redundant comments / redundant doxygen comments. It does not provide valuable information.

carlescufi replied:
But what is the issue with having those comments? They don't look that bad to me

Rafal-Nordic replied:
removed the most controversial comments

jfischer-no commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -0,0 +1,378 @@
+
+int32_t dvfs_service_handler_change_freq_setting(enum dvfs_frequency_setting freq_setting)
+{
+	if (dvfs_service_handler_init_done()) {
+		if (!dvfs_service_handler_freq_change_in_progress()) {
+			if (dvfs_service_handler_freq_setting_allowed(freq_setting)) {
+				nrfs_err_t status =
+					nrfs_dvfs_oppoint_request(freq_setting, get_next_context());
+
+				dvfs_service_handler_nrfs_error_check(status);
+				return status;
+			}
+			return -ENXIO;
+		}
+
+		LOG_INF("Frequency change in progress.");
+		return -EBUSY;
+	}
+
+	LOG_INF("Init not done!");
+	return -EAGAIN;
+}
```
```c

	if (!dvfs_service_handler_init_done()) {
		LOG_INF("Init not done!");
		return -EAGAIN;
	}

	if (dvfs_service_handler_freq_change_in_progress()) {
		LOG_INF("Frequency change in progress.");
		return -EBUSY;
	}

	if (!dvfs_service_handler_freq_setting_allowed(freq_setting)) {
		return -ENXIO;
	}
...
```

jfischer-no commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -0,0 +1,378 @@
+		/* perform background processing */
+		ld_dvfs_scaling_background_process(true);
+	}
+}
+K_THREAD_DEFINE(dvfs_service_handler_task_id,
```
```c
}

K_THREAD_DEFINE(dvfs_service_handler_task_id,
```

Rafal-Nordic commented on modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.h:
```diff
@@ -0,0 +1,94 @@
+enum ipc_channel_status {
+	not_connected = 0,
+	connected     = 1
+};
```
@anangl I had to restore to header file these struct and enum because they are also used in the scfw.

jfischer-no commented on modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.h:
```diff
@@ -0,0 +1,93 @@
+extern "C" {
+#endif
+
+struct __NRFS_PACKED ipc_data_packet_t {
```
Can we use `__packed` here?

`struct __packed ipc_data_packet {`

Please remove ~~_t~~

jfischer-no commented on modules/hal_nordic/nrfs/backends/nrfs_backend_ipc_service.c:
```diff
@@ -0,0 +1,226 @@
+
+#define IPC_INIT_DONE_EVENT (0x01)
+
+struct ipc_channel_config_t {
```
`struct ipc_channel_config {`

~~_t~~

jfischer-no commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -0,0 +1,324 @@
+	}
+}
+
+/**
+ * @brief Task to handle dvfs init procedure.
+ *
+ * @param dummy0
+ * @param dummy1
+ * @param dummy2
+ */
```
`/* Task to handle dvfs init procedure. */`

jfischer-no commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -0,0 +1,324 @@
+	current_freq_setting = oppoint_freq;
+}
+
+/**
+ * @brief Function to set hsfll to highest frequency when switched to ABB.
+ *
+ */
```
`/* Function to set hsfll to highest frequency when switched to ABB. */`

jfischer-no commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -0,0 +1,324 @@
+	}
+}
+
+/**
+ * @brief DVFS event handler callback function.
+ *
+ * @param p_evt event to handle
+ * @param context context data
+ */
+void nrfs_dvfs_evt_handler(nrfs_dvfs_evt_t const *p_evt, void *context)
```
`static void nrfs_dvfs_evt_handler(nrfs_dvfs_evt_t const *p_evt, void *context)`

jfischer-no commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -0,0 +1,324 @@
+	return false;
+}
+
+/**
+ * @brief Function handling steps for scaling preparation.
+ *
+ * @param oppoint_freq requested frequency oppoint
+ */
```
/* Function handling steps for scaling preparation. */

jfischer-no commented on modules/hal_nordic/nrfs/dvfs/ld_dvfs_handler.c:
```diff
@@ -0,0 +1,324 @@
+	return current_freq_setting;
+}
+
+/**
+ * @brief Function to check if current operation is down-scaling
+ *
+ * @param target_freq_setting frequency setting target
+ * @return true down-scaling procedure
+ * @return false not down-scaling procedure
+ */
```
/* Function to check if current operation is down-scaling */

anangl commented on modules/hal_nordic/nrfs/CMakeLists.txt:
```diff
@@ -0,0 +1,41 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library()
```
Please move this under `if(CONFIG_NRFS)` to prevent the following CMake warning:
```
  No SOURCES given to Zephyr library: modules__hal_nordic__nrfs
```
when NRFS is not enabled.

tmon-nordic commented:
When I disabled `NRFS_DVFS_SERVICE_ENABLED` Kconfig symbol, the build failed with undefined references to  `nrfs_dvfs_init_prepare_request' and `nrfs_dvfs_ready_to_scale'. There is either some missing dependencies in Kconfig symbols or the Kconfigs are simply incorrect (it shouldn't be possible to get broken builds by disabling Kconfig symbols in menuconfig).
